client/game.js

```diff
@@ -87,43 +113,73 @@
     // Networking
     //
 
-    var handleMessage = function(msg) {
+    var handleWelcome = function(msg) {
         // TODO do stuff~
     };
 
-    var sendMessage = function(msg) {
-        msg = JSON.stringify(msg);
-        socket.send(msg);
+    var handleGoAway = function(msg) {
+        alert('Server terminated our connection: ' + msg.reason);
     };
 
-    var connect = function() {
-        console.log('Trying to connect to ' + address);
+    var handlePlayerJoined = function(msg) {
+        // TODO do stuff~
+    };
 
-        socket = new WebSocket(address);
+    var handlePlayerLeft = function(msg) {
+        // TODO do stuff~
+    };
 
-        socket.onopen = function() {
-            console.log('Connected to ' + address);
-            setupGame();
-        };
+    var handleShotsFired = function(msg) {
+        // TODO do stuff~
+    };
 
-        socket.onmessage = function(e) {
-            var message = JSON.parse(e.data);
-            handleMessage(msg);
-        };
+    var handlePlayerSpawned = function(msg) {
+        // TODO do stuff~
+    };
 
-        socket.onerror = socket.onclose = function() {
-            console.log('Disconnected from ' + address);
-            // TODO automatic reconnection
-            stopGame();
-        };
+    var handlePlayerDestroyed = function(msg) {
+        // TODO do stuff~
+    };
+
+    var handlePlayerMoving = function(msg) {
+        // TODO do stuff~
+    };
+
+    var handlePlayerStopped = function(msg) {
+        // TODO do stuff~
+    };
+
+    var handleWorldState = function(msg) {
+        // TODO do stuff~
+    };
+
+    var connect = function(address) {
+        transport = new Game.WSTransport(address);
+        transport.addListener('connect',    setupGame);
+        transport.addListener('disconnect', stopGame);
+
+        socket = new Game.Socket(transport);
+        socket.addListener('welcome',          handleWelcome);
+        socket.addListener('go_away',          handleGoAway);
+        socket.addListener('player_joined',    handlePlayerJoined);
+        socket.addListener('player_left',      handlePlayerLeft);
+        socket.addListener('shots_fired',      handleShotsFired);
+        socket.addListener('player_spawned',   handlePlayerSpawned);
+        socket.addListener('player_destroyed', handlePlayerDestroyed);
+        socket.addListener('player_moving',    handlePlayerMoving);
+        socket.addListener('player_stopped',   handlePlayerStopped);
+        socket.addListener('world_state',      handleWorldState);
+
+        transport.connect();
     };
 
     var disconnect = function() {
-        if (socket !== null) {
-            console.log('Disconnecting from ' + address);
-            socket.close();
-            socket = null;
+        if (transport !== null) {
+            transport.disconnect();
```

Comment:
Perhaps we should send a message stating the we are deliberately disconnecting, could help us track down if we have issues with random drops.